esiclient/v1/port_forward.py

```diff
@@ -0,0 +1,422 @@
+            raise ValueError("invalid subnet name")
+        return subnet
+
+    def __repr__(self):
+        return "subnet name or id"
+
+
+class NetworkOpsMixin:
+    def find_floating_ip(self, parsed_args):
+        connection = self.app.client_manager.sdk_connection
+        if isinstance(
+            parsed_args.external_ip, (ipaddress.IPv4Address, ipaddress.IPv6Address)
+        ):
+            # we were given an ip address, so find the matching floating ip
+            fip = connection.network.find_ip(str(parsed_args.external_ip))
+            if fip is None:
+                raise KeyError(f"unable to find floating ip {parsed_args.external_ip}")
+            return fip
+
+        raise ValueError("invalid external ip address")
+
+    def find_or_create_floating_ip(self, parsed_args):
+        connection = self.app.client_manager.sdk_connection
+        try:
+            return self.find_floating_ip(parsed_args)
+        except ValueError:
+            # we were given a network, so attempt to create a floating ip
+            fip = connection.network.create_ip(
+                floating_network_id=parsed_args.external_ip.id,
+            )
+
+        return fip
+
+    def find_port(self, parsed_args):
+        connection = self.app.client_manager.sdk_connection
+        if isinstance(
+            parsed_args.internal_ip, (ipaddress.IPv4Address, ipaddress.IPv6Address)
+        ):
+            # see if there exists a port with the given internal ip
+            ports = list(
+                connection.network.ports(
+                    fixed_ips=f"ip_address={parsed_args.internal_ip}"
+                )
+            )
+
+            # error out if we find multiple matches
+            if len(ports) > 1:
+                raise ValueError(
+                    f"found multiple ports matching address {parsed_args.internal_ip}"
+                )
+
+            # if there was a single port, use it
+            if len(ports) == 1:
+                return ports[0]
+
+            raise KeyError(
+                f"unable to find port with address {parsed_args.internal_ip}"
+            )
+        else:
+            # we already have a port, so just return it
+            return parsed_args.internal_ip
+
+    def find_or_create_port(self, parsed_args):
+        connection = self.app.client_manager.sdk_connection
+        try:
+            return self.find_port(parsed_args)
+        except KeyError:
+            # we need to create a port, which means we need to know the appropriate internal network
+            if parsed_args.internal_ip_network is None:
+                if parsed_args.internal_ip_subnet is None:
+                    raise ValueError(
+                        "unable to create a port because --internal-ip-network is unset"
+                    )
+                internal_network_id = parsed_args.internal_ip_subnet.network_id
+            else:
+                internal_network_id = parsed_args.internal_ip_network.id
```

Comment:
Reading this section, you don't actually use the network/subnet objects set in `parsed_args.internal_ip_subnet` or `parsed_args.internal_ip_network`; you just grab their UUID. Do the API calls below take a name-or-UUID identifier? If so, would it make sense to avoid the extra API calls to grab the subnet/network, just use the user-provided identifier strings, and allow those API calls to error out if the matching subnet/network can't be found?

Reply:
I am a big fan of validating option arguments as early as possible. This avoids situation where you create some resources (say, a floating ip) and then error out afterwards before completing the port forward configuration.

Reply:
fair enough - I'd suggest revisiting this topic if this command takes a long time to run, but I don't expect this command to be spammed.